File: EndpointSecurity/xpc/EventsUI/EventsUI/C/Client/XpcClient.m
//
//  XpcClient.m
//  EventsUI
//
//  Created by msi on 2023/9/9.
//

#import "XpcClient.h"
#import "EService.h"
#import "swr.h"

@interface XpcClient () <EServiceClient>

@property(nonatomic, strong)NSXPCConnection *connection;
@property(nonatomic, strong)NSMutableArray *messages;

@end

@implementation XpcClient

+(instancetype)share {
    static XpcClient *client = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        client = [[XpcClient alloc]init];
    });
    return client;
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.messages = [NSMutableArray array];
    }
    return self;
}

#pragma mark - 方法

- (NSError*)clearCache:(nonnull void (^)(es_clear_cache_result_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy clearCache:callback];
    }
    return resError;
}

- (NSError*)invertMuting:(es_mute_inversion_type_t)type callback:(nonnull void (^)(es_return_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy invertMuting:type callback:callback];
    }
    return resError;
}

- (NSError*)muteProcess:(audit_token_t)audit_token callback:(nonnull void (^)(es_return_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy muteProcess:audit_token callback:callback];
    }
    return resError;
}

- (NSError*)muteProcess:(audit_token_t)audit_token events:(nonnull NSData *)events callback:(nonnull void (^)(es_return_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy muteProcess:audit_token events:events callback:callback];
    }
    return resError;
}

- (NSError*)mutedPath:(nonnull NSString *)path type:(es_mute_path_type_t)type callback:(nonnull void (^)(es_return_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy mutedPath:path type:type callback:callback];
    }
    return resError;
}

- (NSError*)mutedPath:(nonnull NSString *)path type:(es_mute_path_type_t)type events:(nonnull NSData *)events callback:(nonnull void (^)(es_return_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy mutedPath:path type:type events:events callback:callback];
    }
    return resError;
}

- (NSError*)mutedPathsEvents:(nonnull void (^)(es_return_t, NSData * _Nonnull))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy mutedPathsEvents:callback];
    }
    return resError;
}

- (NSError*)mutedProcessesEvents:(nonnull void (^)(es_return_t, NSData * _Nonnull))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy mutedProcessesEvents:callback];
    }
    return resError;
}

- (NSError*)mutingInverted:(es_mute_inversion_type_t)type callback:(nonnull void (^)(es_mute_inverted_return_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy mutingInverted:type callback:callback];
    }
    return resError;
}

- (NSError*)subScriptions:(nonnull void (^)(es_return_t, NSData * _Nonnull))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy subScriptions:callback];
    }
    return resError;
}

- (NSError*)subscribe:(nonnull NSData *)types callback:(nonnull void (^)(es_return_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy subscribe:types callback:callback];
    }
    return resError;
}

- (NSError*)unmutePath:(nonnull NSString *)path type:(es_mute_path_type_t)type callback:(nonnull void (^)(es_return_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy unmutePath:path type:type callback:callback];
    }
    return resError;
}

- (NSError*)unmutePath:(nonnull NSString *)path type:(es_mute_path_type_t)type events:(nonnull NSData *)events callback:(nonnull void (^)(es_return_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy unmutePath:path type:type events:events callback:callback];
    }
    return resError;
}

- (NSError*)unmuteProcess:(audit_token_t)audit_token callback:(nonnull void (^)(es_return_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy unmuteProcess:audit_token callback:callback];
    }
    return resError;
}

- (NSError*)unmutedAllPathsCallback:(nonnull void (^)(es_return_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy unmutedAllPathsCallback:callback];
    }
    return resError;
}

- (NSError*)unmutedAllTargetPathsCallback:(nonnull void (^)(es_return_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy unmutedAllTargetPathsCallback:callback];
    }
    return resError;
}

- (NSError*)unsubscribe:(nonnull NSData *)types callback:(nonnull void (^)(es_return_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy unsubscribe:types callback:callback];
    }
    return resError;
}

- (NSError*)unsubscribeAll:(nonnull void (^)(es_return_t))callback {
    if (!self.connection) {
        return [NSError errorWithDomain:[NSString stringWithFormat:@"%@:%@:%d", [NSString stringWithUTF8String:__FILE__],[NSString stringWithUTF8String:__FUNCTION__], __LINE__] code:1 userInfo:nil];
    }
    __block NSError * resError = nil;
    id<EService> proxy = [self.connection synchronousRemoteObjectProxyWithErrorHandler:^(NSError * _Nonnull error) {
        resError = error;
    }];
    if (!resError) {
        [proxy unsubscribeAll:callback];
    }
    return resError;
}

#pragma - mark EServiceClient

- (void)onType:(ESCType)type data:(nonnull NSData *)data {
    switch (type) {
        case ESCTypeMessage:{
            es_message_t * msg = malloc(sizeof(es_message_t));
            swr_es_message_t_read(msg, (void *)data.bytes);
            @synchronized (self.messages) {
                
            }
        }break;
        default:{
            
        }break;
    }
}

@end

File: EndpointSecurity/xpc/EventsUI/EventsUI/Client/XpcClient.swift
//
//  XpcClient.swift
//  EventsUI
//
//  Created by lyf on 2023/9/8.
//

import Foundation
import EService
import Swr

class XpcClient: NSObject {
    static let share = XpcClient()
    
    var connection: NSXPCConnection!
    
    func connectService(name: String) {
        self.connection = NSXPCConnection(machServiceName: name)
        self.connection.remoteObjectInterface = NSXPCInterface(with: EService.self)
        self.connection.exportedInterface = NSXPCInterface(with:EServiceClient.self)
        self.connection.exportedObject = self;
        self.connection.resume()
    }

}

extension XpcClient: EServiceClient {
    func onType(_ type: ESCType, data: Data) {
        switch type {
        case ESCTypeMessage:
            var msg = malloc(MemoryLayout<es_message_t>.size).bindMemory(to: es_message_t.self, capacity: 1)
            data.by
            swr_es_message_t_read(msg, (data as NSData).bytes)
            print(data)
        default:
            print(data)
        }
    }
}

extension XpcClient {
    
    func subscribe(_ types: Data, callback: @escaping (es_return_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.subscribe(types, callback: callback);
        }
        return err
    }
    
    func unsubscribe(_ types: Data, callback: @escaping (es_return_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.unsubscribe(types, callback: callback);
        }
        return err
    }
    
    func unsubscribeAll(_ callback: @escaping (es_return_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.unsubscribeAll(callback);
        }
        return err
    }
    
    func subScriptions(_ callback: @escaping (es_return_t, Data) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.subScriptions(callback);
        }
        return err
    }
    
    func muteProcess(_ audit_token: audit_token_t, callback: @escaping (es_return_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.muteProcess(audit_token, callback: callback);
        }
        return err
    }
    
    func unmuteProcess(_ audit_token: audit_token_t, callback: @escaping (es_return_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.unmuteProcess(audit_token, callback: callback);
        }
        return err
    }
    
    func muteProcess(_ audit_token: audit_token_t, envents types: Data, callback: @escaping (es_return_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.muteProcess(audit_token, envents: types, callback: callback);
        }
        return err
    }
    
    func mutedProcessesEvents(_ callback: @escaping (es_return_t, Data) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.mutedProcessesEvents(callback);
        }
        return err
    }
    
    func mutedPath(_ path: String, type: es_mute_path_type_t, callback: @escaping (es_return_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.mutedPath(path, type: type, callback: callback);
        }
        return err
    }
    
    func mutedPath(_ path: String, type: es_mute_path_type_t, events: Data, callback: @escaping (es_return_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.mutedPath(path, type: type, events: events, callback: callback);
        }
        return err
    }
    
    func unmutedAllPathsCallback(_ callback: @escaping (es_return_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.unmutedAllPathsCallback(callback);
        }
        return err
    }
    
    func unmutedAllTargetPathsCallback(_ callback: @escaping (es_return_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.unmutedAllTargetPathsCallback(callback);
        }
        return err
    }
    
    func unmutePath(_ path: String, type: es_mute_path_type_t, callback: @escaping (es_return_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.unmutePath(path, type: type, callback: callback);
        }
        return err
    }
    
    func unmutePath(_ path: String, type: es_mute_path_type_t, events: Data, callback: @escaping (es_return_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.unmutePath(path, type: type, events: events, callback: callback);
        }
        return err
    }
    
    func mutedPathsEvents(_ callback: @escaping (es_return_t, Data) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.mutedPathsEvents(callback);
        }
        return err
    }
    
    func invertMuting(_ type: es_mute_inversion_type_t, callback: @escaping (es_return_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.invertMuting(type, callback: callback);
        }
        return err
    }
    
    func mutingInverted(_ type: es_mute_inversion_type_t, callback: @escaping (es_mute_inverted_return_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.mutingInverted(type, callback: callback);
        }
        return err
    }
    
    func clearCache(_ callback: @escaping (es_clear_cache_result_t) -> Void) -> Error? {
        var err: Error? = nil
        if let proxy = self.connection.synchronousRemoteObjectProxyWithErrorHandler({ error in
            err = error
        }) as? EService {
            proxy.clearCache(callback);
        }
        return err
    }
}
